pkg/collect/host_cgroup_linux.go

```diff
@@ -0,0 +1,321 @@
+// This Control Groups collector is heavily based on k0s'
+// probes implementation https://github.com/k0sproject/k0s/blob/main/internal/pkg/sysinfo/probes/linux/cgroups.go
+
+//go:build linux
+
+package collect
+
+import (
+	"errors"
+	"fmt"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"sort"
+	"strings"
+	"syscall"
+
+	"github.com/cilium/ebpf/rlimit"
+	"github.com/containerd/cgroups/v3/cgroup2"
+	"github.com/opencontainers/runtime-spec/specs-go"
+	"golang.org/x/sys/unix"
+
+	"k8s.io/klog/v2"
+	"k8s.io/utils/ptr"
+)
+
+func discoverConfiguration(mountPoint string) (cgroupsResult, error) {
+	results := cgroupsResult{}
+
+	var st syscall.Statfs_t
+	if err := syscall.Statfs(mountPoint, &st); err != nil {
+		if os.IsNotExist(err) {
+			klog.V(2).Infof("no file system mounted at %q", mountPoint)
+			return results, nil
+		}
+
+		return results, fmt.Errorf("failed to stat %q: %w", mountPoint, err)
+	}
+
+	switch st.Type {
+	case unix.CGROUP2_SUPER_MAGIC:
+		klog.V(2).Infof("cgroup v2 mounted at %q", mountPoint)
+		// Discover cgroup2 and controllers enabled
+		// https://www.kernel.org/doc/html/v5.16/admin-guide/cgroup-v2.html#mounting
+		v, err := discoverV2Configuration(mountPoint)
+		if err != nil {
+			return results, err
+		}
+		results.CGroupV2 = v
+	case unix.CGROUP_SUPER_MAGIC, unix.TMPFS_MAGIC:
+		klog.V(2).Infof("cgroup v1 mounted at %q", mountPoint)
+		// Discover cgroup1 and controllers enabled
+		// https://git.kernel.org/pub/scm/docs/man-pages/man-pages.git/tree/man7/cgroups.7?h=man-pages-5.13#n159
+		// https://www.kernel.org/doc/html/v5.16/admin-guide/cgroup-v1/cgroups.html#how-do-i-use-cgroups
+		r, err := discoverV1Configuration(mountPoint)
+		if err != nil {
+			return results, err
+		}
+		results.CGroupV1 = r
+	default:
+		return results, fmt.Errorf("unexpected file system type of %q: 0x%x", mountPoint, st.Type)
+	}
+
+	// If cgroup1 or cgroup2 is enabled
+	results.CGroupEnabled = results.CGroupV1.Enabled || results.CGroupV2.Enabled
+
+	// Sort controllers for consistent output
+	if len(results.CGroupV1.Controllers) > 0 {
+		sort.Strings(results.CGroupV1.Controllers)
+	} else {
+		results.CGroupV1.Controllers = []string{}
+	}
+	if len(results.CGroupV2.Controllers) > 0 {
+		sort.Strings(results.CGroupV1.Controllers)
+	} else {
+		results.CGroupV2.Controllers = []string{}
+	}
+
+	// Combine all controllers
+	set := make(map[string]struct{})
+	for _, c := range results.CGroupV1.Controllers {
+		set[c] = struct{}{}
+	}
+
+	for _, c := range results.CGroupV2.Controllers {
+		set[c] = struct{}{}
+	}
+
+	for c := range set {
+		results.AllControllers = append(results.AllControllers, c)
+	}
+	sort.Strings(results.AllControllers)
+
+	return results, nil
+}
+
+func discoverV1Configuration(mountPoint string) (cgroupResult, error) {
+	res := cgroupResult{}
+	// Get the available controllers from /proc/cgroups.
+	// See https://www.man7.org/linux/man-pages/man7/cgroups.7.html#NOTES
+
+	f, err := os.Open("/proc/cgroups")
```

Comment:
For cgroup v1, information about controllers is always stored in `/proc/cgroups` file. This is different from the cgroup mount points.

From https://www.man7.org/linux/man-pages/man7/cgroups.7.html#NOTES
```
   /proc files
       /proc/cgroups (since Linux 2.6.24)
              This file contains information about the controllers that
              are compiled into the kernel.  An example of the contents
              of this file (reformatted for readability) is the
```